### PR TITLE
[fix][tiered-storage] Don't cleanup data when offload met Metastore exception

### PR DESCRIPTION
**Pull request by zymap**
---

### Motivation

There have two ways that will cause the offload data cleanup. One is met offload conflict exception, and another is completeLedgerInfoForOffloaded reaches max retry time and throws zookeeper exceptions.

We retry the zookeeper operation on connection loss exception. We should be careful about this exception, because we may lose data if the metadata updated successfully.

When a MetaStore exception happens, we can not make sure the metadata update is failed or not. Because we have a retry on the connection loss, it is possible to get a BadVersion or other exception after retrying.

So we don't clean up the data if this happens.

### Modifications

- don't delete data if has a meta store exception

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

**Review by Technoboy- — APPROVED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 1]
```diff
@@ -3006,6 +3006,13 @@ private void offloadLoop(CompletableFuture<PositionImpl> promise, Queue<LedgerIn
                                     if (exception != null) {
                                         log.error("[{}] Failed to offload data for the ledgerId {}",
                                                 name, ledgerId, exception);
+                                        if (exception instanceof MetaStoreException) {
+                                            // When a MetaStore exception happens, we can not make sure the metadata
+                                            // update is failed or not. Because we have a retry on the connection loss,
+                                            // it is possible to get a BadVersion or other exception after retrying.
+                                            // So we don't clean up the data if it has metadata operation exception.
```
Comment (eolivelli) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:3013]: we must log a (ERROR) message, this way we can see this case in the logs

**Review by eolivelli — CHANGES_REQUESTED** [review]
LGTM
but we must log something that explains what happened

**Comment by zymap** [comment]
@eolivelli Added. PTAL

**Review by eolivelli — APPROVED** [review]
LGTM

**Comment by hangc0276** [comment]
Nice catch!

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 2]
```diff
@@ -3004,8 +3004,19 @@ private void offloadLoop(CompletableFuture<PositionImpl> promise, Queue<LedgerIn
...
                             .whenComplete((ignore2, exception) -> {
                                     if (exception != null) {
-                                        log.error("[{}] Failed to offload data for the ledgerId {}",
+                                        Throwable e = FutureUtil.unwrapCompletionException(exception);
+                                        if (e instanceof MetaStoreException) {
+                                            // When a MetaStore exception happens, we can not make sure the metadata
+                                            // update is failed or not. Because we have a retry on the connection loss,
+                                            // it is possible to get a BadVersion or other exception after retrying.
+                                            // So we don't clean up the data if it has metadata operation exception.
+                                            log.error("[{}] Failed to update offloaded metadata for the ledgerId {}",
```
Comment (codelipenghui) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:3013]: I think it's better to add more information like

`Failed to update offloaded metadata for the ledgerId, the offloaded data will not be cleaned up`

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 3]
```diff
@@ -3004,8 +3004,19 @@ private void offloadLoop(CompletableFuture<PositionImpl> promise, Queue<LedgerIn
...
                                                 name, ledgerId, exception);
+                                            return;
+                                        } else {
+                                            log.error("[{}] Failed to offload data for the ledgerId {}",
+                                                name, ledgerId, exception);
```
Comment (codelipenghui) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:3018]: `Failed to offload data for the ledgerId, try to clean up the offloaded data`

**Review by codelipenghui — COMMENTED** [review]
I have left 2 minor comments about the logs.
Provide more information which can help people like operation team who don't have such detailed knowledge to understand the behavior.

**Comment by zymap** [comment]
@codelipenghui Done. PTAL, thanks

**Pull request merged**